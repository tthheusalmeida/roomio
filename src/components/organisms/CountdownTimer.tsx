import { useEffect, useState } from "react";
import { mergeClassNames } from "@/utils/classNames";

interface CountdownTimerProps {
  initialMinutes: number;
  onFinish?: () => void;
  onProgress?: (progress: number) => void;
  className?: string;
}

export default function CountdownTimer({
  initialMinutes,
  onFinish,
  onProgress,
  className,
}: CountdownTimerProps) {
  const totalSeconds = initialMinutes * 60;
  const [seconds, setSeconds] = useState(totalSeconds);

  useEffect(() => {
    if (seconds <= 0) {
      onFinish?.();
      return;
    }

    const timer = setInterval(() => {
      setSeconds((prev) => prev - 1);
    }, 1000);

    return () => clearInterval(timer);
  }, [seconds, onFinish]);

  useEffect(() => {
    if (onProgress) {
      onProgress(((totalSeconds - seconds) / totalSeconds) * 100);
    }
  }, [seconds, totalSeconds, onProgress]);

  return (
    <div className={mergeClassNames(className)}>
      <p>{formatTime(seconds)}</p>
    </div>
  );
}

function formatTime(totalSeconds: number): string {
  const roundedSeconds = Math.ceil(totalSeconds);
  const minutes = Math.floor(roundedSeconds / 60);
  const seconds = roundedSeconds % 60;
  return `${String(minutes).padStart(2, "0")}:${String(seconds).padStart(
    2,
    "0"
  )}`;
}
