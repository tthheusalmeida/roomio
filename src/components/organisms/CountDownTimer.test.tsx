import { render, screen, act } from "@testing-library/react";
import CountdownTimer from "./CountdownTimer";

jest.useFakeTimers();

describe("CountdownTimer", () => {
  afterEach(() => {
    jest.clearAllTimers();
  });

  it("renders initial time correctly", () => {
    render(<CountdownTimer initialMinutes={1} />);
    expect(screen.getByText("01:00")).toBeInTheDocument();
  });

  it("calls onProgress as time passes", () => {
    const onProgress = jest.fn();
    render(<CountdownTimer initialMinutes={1} onProgress={onProgress} />);

    act(() => {
      jest.advanceTimersByTime(30000); // 30s
    });

    const lastCall = onProgress.mock.calls[onProgress.mock.calls.length - 1][0];
    expect(lastCall).toBeCloseTo(50, 0);
  });

  it("calls onFinish when countdown reaches zero", () => {
    const onFinish = jest.fn();
    render(<CountdownTimer initialMinutes={0.05} onFinish={onFinish} />); // 3s

    act(() => {
      jest.advanceTimersByTime(4000);
    });

    expect(onFinish).toHaveBeenCalled();
  });

  it("updates displayed time correctly", () => {
    render(<CountdownTimer initialMinutes={0.05} />); // ~3s

    expect(screen.getByText("00:03")).toBeInTheDocument();

    act(() => {
      jest.advanceTimersByTime(1000);
    });
    expect(screen.getByText("00:02")).toBeInTheDocument();

    act(() => {
      jest.advanceTimersByTime(1000);
    });
    expect(screen.getByText("00:01")).toBeInTheDocument();

    act(() => {
      jest.advanceTimersByTime(1000);
    });
    expect(screen.getByText("00:00")).toBeInTheDocument();
  });

  it("matches snapshot", () => {
    const { container } = render(<CountdownTimer initialMinutes={1} />);
    expect(container).toMatchSnapshot();
  });
});
