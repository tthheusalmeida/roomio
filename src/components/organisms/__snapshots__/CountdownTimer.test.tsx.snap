// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`CountdownTimer matches snapshot 1`] = `
<div>
  <div
    class=""
  >
    <p>
      01:00
    </p>
  </div>
</div>
`;
